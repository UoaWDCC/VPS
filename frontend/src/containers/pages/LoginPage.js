import React, { useContext, useEffect } from "react";
import { useHistory } from "react-router-dom";
import AuthenticationContext from "../../context/AuthenticationContext";
import LoadingPage from "./LoadingPage";

export default function LoginPage() {
  const { user } = useContext(AuthenticationContext);
  const history = useHistory();

  useEffect(() => {
    if (user) {
      history.push("/");
    }
  }, [user]);

  return <LoadingPage text="Redirecting..." />;
}
