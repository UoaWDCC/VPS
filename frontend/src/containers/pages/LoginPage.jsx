import { useContext, useEffect } from "react";
import { useHistory } from "react-router-dom";
import AuthenticationContext from "../../context/AuthenticationContext";
import { usePost } from "../../hooks/crudHooks";
import LoadingPage from "./LoadingPage";

/**
 * Container for the login page. Redirects logged-in users to main page and allows users to sign up using Google.
 *
 * @container
 */
export default function LoginPage() {
  const { user, loading, signInUsingGoogle, getUserIdToken } = useContext(
    AuthenticationContext
  );
  const history = useHistory();

  useEffect(() => {
    if (user) {
      usePost(
        `/api/user`,
        {
          name: user.displayName,
          uid: user.uid,
          email: user.email,
          pictureURL: user.photoURL,
        },
        getUserIdToken
      );
      history.push("/");
    }
  }, [user]);

  // useEffect(() => {
  //   if (!loading && !user) {
  //     signInUsingGoogle();
  //   }
  // }, [loading]);

  return (
    <div>
      <button
        type="button"
        onClick={() => {
          if (!loading && !user) {
            signInUsingGoogle();
          } else {
            <LoadingPage text="logging in..." />;
          }
        }}
      >
        WIP Login
      </button>
    </div>
  );
}
