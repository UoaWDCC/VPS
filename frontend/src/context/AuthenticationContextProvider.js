import React, { useState } from "react";
import { signInWithRedirect } from "firebase/auth";
import { useAuthState } from "react-firebase-hooks/auth";
import AuthenticationContext from "./AuthenticationContext";
import { auth, googleProvider } from "../firebase/firebase";
import { useGetSimplified } from "../hooks/crudHooks";

/**
 * This is a Context Provider made with the React Context API
 * AuthenticationContext grants access to functions and variables related to Firebase login
 */
export default function AuthenticationContextProvider({ children }) {
  const [user, loading, error] = useAuthState(auth);

  /**
   * No idToken is stored in state to ensure the non-expired idToken is always used
   * @returns idToken or null if user is not signed in
   */
  async function getUserIdToken() {
    if (user) {
      const token = await user.getIdToken();
      return token;
    }

    return null;
  }

  function signInUsingGoogle() {
    signInWithRedirect(auth, googleProvider);
  }

  function signOut() {
    auth.signOut();
  }

  /**
   * This function determines the role of the current logged in user.
   * @returns the role of user
   */
  function getRole() {
    const [staffList, setStaffList] = useState();
    useGetSimplified("/api/staff", setStaffList);

    let onStaffList = false;
    if (user && getUserIdToken() != null) {
      if (staffList !== undefined) {
        for (let i = 0; i < staffList.length; i += 1) {
          if (staffList[i].firebaseID === user.uid) {
            onStaffList = true;
          }
        }
      }

      if (onStaffList) {
        return "admin";
      }
      return "user";
    }

    return "nonexistent";
  }

  // creating user object with role property
  const VpsUser = {
    firebaseUserObj: user,
    role: getRole(),
  };

  return (
    <AuthenticationContext.Provider
      value={{
        getUserIdToken,
        loading,
        user,
        error,
        signOut,
        signInUsingGoogle,
        VpsUser,
      }}
    >
      {children}
    </AuthenticationContext.Provider>
  );
}
