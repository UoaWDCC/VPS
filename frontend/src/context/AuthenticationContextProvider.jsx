import { signInWithRedirect, signInWithPopup } from "firebase/auth";
import { useState } from "react";
import { useAuthState } from "react-firebase-hooks/auth";
import { auth, googleProvider } from "../firebase/firebase";
import { useGetSimplified } from "../hooks/crudHooks";
import AuthenticationContext from "./AuthenticationContext";

/**
 * This is a Context Provider made with the React Context API
 * AuthenticationContext grants access to functions and variables related to Firebase login
 */
export default function AuthenticationContextProvider({ children }) {
  const [user, loading, error] = useAuthState(auth);

  /**
   * No idToken is stored in state to ensure the non-expired idToken is always used
   * @returns idToken or null if user is not signed in
   */
  async function getUserIdToken() {
    if (user) {
      const token = await user.getIdToken();
      return token;
    }

    return null;
  }

  function signInUsingGoogle() {
    signInWithPopup(auth, googleProvider);
    // signInWithRedirect(auth, googleProvider);
  }

  function signOut() {
    auth.signOut();
  }

  // getting role from backend
  const [userRole, setUserRole] = useState();
  const userID = user == null ? "null" : user.uid; // this is to avoid null pointer exceptions while confining to hook rules
  useGetSimplified(`/api/staff/${userID}`, setUserRole);

  // creating user object with role property
  const VpsUser = {
    firebaseUserObj: user,
    role: userRole,
  };

  return (
    <AuthenticationContext.Provider
      value={{
        getUserIdToken,
        loading,
        user,
        error,
        signOut,
        signInUsingGoogle,
        VpsUser,
      }}
    >
      {children}
    </AuthenticationContext.Provider>
  );
}
