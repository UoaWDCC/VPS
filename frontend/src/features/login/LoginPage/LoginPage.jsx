import { useContext, useEffect } from "react";
import { useHistory, useLocation } from "react-router-dom";
import AuthenticationContext from "context/AuthenticationContext";
import LoadingPage from "features/status/LoadingPage";
import styles from "./LoginPage.module.scss";
import GoogleIcon from "./GoogleIcon";

import toast from "react-hot-toast";
import axios from "axios";

const handleSignIn = async (user) => {
  const token = await user.getIdToken();
  const config = {
    method: "post",
    url: `api/user/`,
    headers: { Authorization: `Bearer ${token}` },
    data: {
      name: user.displayName,
      uid: user.uid,
      email: user.email,
      pictureURL: user.photoURL,
    },
  };
  return axios.request(config);
};

/**
 * Container for the login page. Redirects logged-in users to main page and allows users to sign up using Google.
 *
 * @container
 */
export default function LoginPage() {
  const { user, loading, isSigningIn, signInUsingGoogle, signOut } = useContext(
    AuthenticationContext
  );
  const history = useHistory();
  const location = useLocation();

  const params = new URLSearchParams(location.search);
  const redirectPath = params.get("redirect") || "/";

  useEffect(() => {
    if (!user) return;

    handleSignIn(user)
      .then(() => history.push(redirectPath))
      .catch((e) => {
        if (e.response?.status === 403) {
          toast.error("Please sign in with your UoA account");
          signOut();
        } else {
          console.log(e);
          toast.error("An unexpected error occurred while signing in");
        }
      });
  }, [user]);

  return (
    <div className="w-full h-full flex flex-col items-center p-10 graph-paper">
      {/* Main title + subtitle */}
      <div className="flex flex-col justify-end items-center basis-[60%] gap-5 text-center font-mona">
        <h1 className={`font-semibold text-6xl text-uoa-blue`}>
          Virtual Patient Simulator
        </h1>
        <p className="text-slate-600 max-w-[65rem]">
          This tool aims to provide Medical and Health Science students at the
          University of Auckland with a tool that supports interactive and
          immersive education through virtual patient scenarios.
        </p>

        <button
          className="btn btn-outline mt-16 cursor-pointer"
          disabled={loading || isSigningIn}
          onClick={() => {
            if (!loading && !user) {
              signInUsingGoogle();
            } else {
              <LoadingPage text="logging in..." />;
            }
          }}
        >
          {loading || isSigningIn ? (
            <>
              <span className="loading loading-spinner"></span> Check the popup!
            </>
          ) : (
            <>
              <GoogleIcon /> Sign In with Google
            </>
          )}
        </button>
        <p className="font-light font-mona text-xs text-slate-400">
          Make sure to use your <b>university account</b>.
        </p>
      </div>

      {/* credits */}
      <div className="w-full font-mona font-light text-xs tracking-wider text-slate-400 flex-grow flex justify-center items-end text-center">
        <p>
          Crafted by many hands, across many teams, over many years at{" "}
          <a
            className={`font-bold ${styles.gradient} `}
            href="https://wdcc.co.nz/"
            target="_blank"
          >
            WDCC
          </a>
          .
        </p>
      </div>

      {/* Hidden on small screens. UoA is not *that* important in this... */}
      <a href="https://www.auckland.ac.nz" target="_blank">
        <img
          draggable="false"
          src="uoa-logo-filled.png"
          alt="University of Auckland Logo"
          className="w-10 right-7 bottom-7 fixed md:block hidden"
        />
      </a>
    </div>
  );
}
