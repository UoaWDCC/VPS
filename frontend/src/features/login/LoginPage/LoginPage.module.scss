.gradient {
  background-image: linear-gradient(
    to right,
    #6366f1 10%,
    /* indigo-500 */ #0ea5e9 30%,
    /* sky-500 */ #10b981 90% /* emerald-500 */
  );
  background-clip: text;
  color: transparent;
  -webkit-background-clip: text; /* For better support in webkit-based browsers */
}
