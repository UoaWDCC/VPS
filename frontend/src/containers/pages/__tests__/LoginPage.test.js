import { render } from "@testing-library/react";
import { BrowserRouter } from "react-router-dom";
import AuthenticationContextProvider from "../../../context/AuthenticationContextProvider";
import LoginPage from "../LoginPage";

test("Login page snapshot test", () => {
  const { baseElement } = render(
    <BrowserRouter>
      <AuthenticationContextProvider>
        <LoginPage />
      </AuthenticationContextProvider>
    </BrowserRouter>
  );

  expect(baseElement).toMatchSnapshot();
});
