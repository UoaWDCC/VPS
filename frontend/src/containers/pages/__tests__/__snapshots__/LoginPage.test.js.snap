// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login page snapshot test 1`] = `
<body>
  <div>
    <div
      class="rowContainer"
    >
      <div
        class="sideBar"
      >
        <img
          alt="University of Auckland Logo"
          class="logo"
          draggable="false"
          src="uoa-logo.png"
        />
        <div
          class="sideBarText"
        >
          Please login to start your simulation
        </div>
        <button
          class="googleButton"
          type="button"
        >
          <svg
            fill="none"
            height="24"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M5.26644 9.76453C6.19903 6.93863 8.85469 4.90909 12.0002 4.90909C13.6912 4.90909 15.2184 5.50909 16.4184 6.49091L19.9093 3C17.7821 1.14545 15.0548 0 12.0002 0C7.27031 0 3.19799 2.6983 1.24023 6.65002L5.26644 9.76453Z"
              fill="#EA4335"
            />
            <path
              d="M16.0406 18.0142C14.9508 18.718 13.5659 19.0926 11.9998 19.0926C8.86633 19.0926 6.21896 17.0785 5.27682 14.2695L1.2373 17.3366C3.19263 21.2953 7.26484 24.0017 11.9998 24.0017C14.9327 24.0017 17.7352 22.959 19.834 21.0012L16.0406 18.0142Z"
              fill="#34A853"
            />
            <path
              d="M19.8342 20.9978C22.0292 18.9503 23.4545 15.9019 23.4545 11.9982C23.4545 11.2891 23.3455 10.5255 23.1818 9.81641H12V14.4528H18.4364C18.1188 16.0119 17.2663 17.2194 16.0407 18.0108L19.8342 20.9978Z"
              fill="#4A90E2"
            />
            <path
              d="M5.27698 14.2663C5.03833 13.5547 4.90909 12.7922 4.90909 11.9984C4.90909 11.2167 5.03444 10.4652 5.2662 9.76294L1.23999 6.64844C0.436587 8.25884 0 10.0738 0 11.9984C0 13.918 0.444781 15.7286 1.23746 17.3334L5.27698 14.2663Z"
              fill="#FBBC05"
            />
          </svg>
          Sign In with Google
           
        </button>
        <div
          class="sponsorText"
        >
          Developed by UoA and WDCC
        </div>
      </div>
      <div
        class="login"
      >
        <div
          class="header"
        >
          Virtual Patient Simulator
        </div>
        <div
          class="infoBlock"
        >
          This tool aims to provide Medical and Health Science students at the University of Auckland with a tool that supports interactive and immersive education through virtual patient scenarios.
        </div>
      </div>
    </div>
  </div>
</body>
`;
